Fixes #12834 - Avoid select2 integration test deadlock

dLobatog opened the pull request.

dLobatog committed:
Fixes #12834 - Avoid select2 integration test deadlock

The test was testing that selecting a hostgroup (1), then clicking on
inherit environment, then selecting another hostgroup (2) should return the
original environment. That overrides hostgroup 2 environment.

dLobatog commented on test/integration/host_test.rb:
```diff
@@ -118,12 +118,11 @@ def class_params
-
       wait_for_ajax
 
       find("#host_environment_id + .input-group-btn .btn").click
```
I could reproduce it a bit more frequently (1/4 times) by setting a `sleep 2` before the find.

domcleal commented:
Running:
- http://ci.theforeman.org/job/test_develop_pr_core/7413/
- http://ci.theforeman.org/job/test_develop_pr_core/7414/
- http://ci.theforeman.org/job/test_develop_pr_core/7415/
- http://ci.theforeman.org/job/test_develop_pr_core/7416/
- http://ci.theforeman.org/job/test_develop_pr_core/7417/

(and http://ci.theforeman.org/job/test_develop_pull_request/13770/ for the PR generally.)

domcleal commented:
Please check http://ci.theforeman.org/job/test_develop_pr_core/7413/, it's failing an assertion instead now.

domcleal commented:
The assertion is failing because the environment is changed when the host group is changed, even though the inherit button was unpressed.  (It goes back to inherit.)

I noticed this when reviewing 77b36c9, and didn't realise this test specifically tried to test that scenario and expected it to work that way.  The behaviour change in that commit and the preceding one only affected _new_ hosts though, so editing a host should still show the original behaviour that the test expects.

I thought the behaviour change was OK and after a while trying to decide if it was a bug or not, I decided not and didn't comment on it.  If it's OK then we should perhaps change the test to test editing a host instead.

dLobatog commented:
@domcleal Updated, now the test tries to override the hostgroup environment within an 'edit' context vs 'new' as it was before. I also extracted a `click_on_inherit` method and moved the test helper methods to be private.

dLobatog commented:
I've triggered 

http://ci.theforeman.org/job/test_develop_pr_core/7420/
and 
http://ci.theforeman.org/job/test_develop_pr_core/7419/

since the test_develop job is not stable and won't be triggered automatically.

dLobatog commented:
:ok: , both passed. One last sanity check...

http://ci.theforeman.org/job/test_develop_pr_core/7421/

ohadlevy commented:
@ShimShtein mind having a look?

dLobatog commented on test/integration/host_test.rb:
```diff
@@ -533,6 +475,29 @@ def test_show_action_buttons
+      # Clear host environment, then inherit it from hostgroup
+      @host.environment = nil
+      @host.hostgroup = original_hostgroup
+      @host.send(:assign_hostgroup_attributes, ['environment_id'])
```
Just assigning a hostgroup then saving won't make it inherit the hostgroup, as per [this method](https://github.com/theforeman/foreman/blob/develop/app/models/host/managed.rb#L616). Bug?

ShimShtein replied:
If you want to reassign environment from hostgroup, you need to do it differently:

``` Ruby
@host.attributes = @host.apply_inherited_attributes('hostgroup_id' => original_hostgroup.id)
```

domcleal replied:
Or just assign the environment (instead of nil) and host group, skip using this private method.  There's no extra information stored when inheriting, it only sets the environment association.

dLobatog commented on test/integration/host_test.rb:
```diff
@@ -533,6 +475,29 @@ def test_show_action_buttons
+      overridden_hostgroup = FactoryGirl.
+        create(:hostgroup, :environment => FactoryGirl.create(:environment))
+
+      visit edit_host_path(@host)
```
This is basically the only change in behavior with respect to how the test was before. I moved it to the 'edit page' context.

ShimShtein replied:
I think moving to edit will not test the right behavior.
We have different code paths in JS for edited and new hosts: https://github.com/theforeman/foreman/blob/develop/app/assets/javascripts/host_edit.js#L218. Only new hosts will call update_form (and reset the values from server).

domcleal replied:
I'm unsure then from the test title what it should be testing.  It seems to test only the host group change behaviour, not really the override (which wouldn't require a hg change).

Perhaps you can reply to my earlier comment about this then, to say whether the behaviour change in 77b36c9 in the new host case was intended or not.  The change seems OK to me, but the test expectation would need changing if it was to stay against the new host.  Perhaps we should have both, with different expectations - or the behaviour change is a bug, and we need both with the same expectation.

ShimShtein replied:
I think I have the expectations figured now:

For new hosts:
Each change of the hostgroup should reset the environment. Reason: the user chooses how to create a host, so if he is choosing a hostgroup, previous changes to environment are irrelevant.

For edited hosts:
Changes to hostgroup should not affect environment, unless the user requested inheritance explicitly. Reason: the host is already alive, each change should be validated by the user.

domcleal replied:
Indeed, I think that matches my comment at https://github.com/theforeman/foreman/pull/3001#issuecomment-167083945.

Perhaps this PR should keep the "new" test then, but without expecting the original environment to change rather than moving it to "edit" - or even better, keep both.

dLobatog commented on test/integration/host_test.rb:
```diff
@@ -533,6 +475,29 @@ def test_show_action_buttons
+      visit edit_host_path(@host)
+      click_on_inherit('environment')
+      select2(overridden_hostgroup.name, :from => 'host_hostgroup_id')
+      wait_for_ajax
```
This wait_for_ajax was missing before

ShimShtein replied:
:+1: for this one, nice catch!

dLobatog commented:
All 3 Jenkins jobs before are :green_heart:

dLobatog commented:
For context:

#### Why does this deadlock happen?

When the test clicks on select2 to change the hostgroup (for the 2nd time) the browser thread will try to acquire a _share lock_ on smart_proxies_features. At the same time the server thread is trying to setup (or cleanup?) fixtures with an _exclusive lock_.
2 threads trying to acquire a lock (one of the locks is exclusive) over the same resource, thus a deadlock happens. 

You can reproduce it a bit more often by setting 'sleep x' before the `find` method that clicks on inherit. The 2nd 'wait_for_ajax' after the 2nd select2 click fixes the deadlock. but it breaks the original test.

#### Why does the original test fail when the deadlock is fixed by waiting on ajax?

The original test only worked because it didn't wait for the select2 to set attributesonly worked because it didn't wait for the 2nd select2 to set any attributes.

On new hosts, selecting a 2nd hostgroup will override 'environment' and set it to the new hostgroup inherited environment. This behavior is correct IMO as it respects the 'inherit' button. However, the original test was checking that the environment remained the same.

On edited hosts, the behavior is different, selecting a new hostgroup will respect the original choice of environment even if inherit environment is checked. It will uncheck `inherit` and leave the environment as the environment of the original host group. 

I'm not sure whether the 'edit' behavior is wrong or right - on the one hand it protects the user from changing the environment from an already created host, on the other it doesn't respect the hostgroup environment inheritance.

domcleal commented:
> On new hosts, selecting a 2nd hostgroup will override 'environment' and set it to the new hostgroup inherited environment. This behavior is correct IMO as it respects the 'inherit' button. However, the original test was checking that the environment remained the same.

The host group change doesn't respect the inherit button (if it's "unpressed", i.e. not inheriting), it gets reset to an inherit status on hg change.  Before 77b36c9 it left it uninherited on hg change.  I'm fine with either behaviour.

> On edited hosts, the behavior is different, selecting a new hostgroup will respect the original choice of environment even if inherit environment is checked. It will uncheck inherit and leave the environment as the environment of the original host group.
> 
> I'm not sure whether the 'edit' behavior is wrong or right - on the one hand it protects the user from changing the environment from an already created host, on the other it doesn't respect the hostgroup environment inheritance.

This is intended, to minimise changing lots of attrs on an existing host when the hg is changed - have the user opt in to any further changes they want to make without wiping all of their settings.  (As the other way around there wouldn't be an easy way to undo the damage.)

ShimShtein commented:
@domcleal  :+1: For the behavior description, that's exactly my thoughts, look at my comment for your previous question.

@dLobatog My guess is that the deadlock was caused by the fact that the test was finishing before the AJAX call (caused by the second select2 click) finished.

Maybe it will be a good thing to put `wait_for_ajax` at `teardown` for every integration test.

dLobatog commented:
@domcleal @ShimShtein Updated it - now there's one test for 'new host', which checks the environment is always the environment of the hostgroup you choose, and a test for 'edit host' which checks it doesn't override environment when you change the hostgroup.

dLobatog commented:
http://ci.theforeman.org/job/test_develop_pr_core/7423/
http://ci.theforeman.org/job/test_develop_pr_core/7422/

dLobatog commented:
http://ci.theforeman.org/job/test_develop_pr_core/7422/database=postgresql,label=fast,ruby=2.1/console one of the previous jobs didn't fully pass for some other error. 

All other tests passed, however I started another one:
http://ci.theforeman.org/job/test_develop_pr_core/7424/

domcleal commented:
Merged as e47927e, thanks @dLobatog!

PR tests should resume within an hour or two, but there will be a lot to run, so please be patient and wait for it all to clear.
